# Remove repetitions from FASTA file dowloaded from PFAM. 

## When downloading a whole family of proteins sequences from PFAM, some sequences happen to be repeated. In this program we get rid of them.

1. This python scirpt reads a fasta file named `output_file_no_rep.txt` in the same directory as the Python file `Rename_acc_num_source_org.py`.
2. This Fasta file contains all protein sequences from GH68 family of bacteria dowloaded from PFAM (https://pfam.xfam.org/family/PF02435#tabview=tab7.
3. The Fasta file should contain the Accession Number followed by the sequence in the text like in the following example: 

```
>A0A3M8AKQ8.1/53-513 Glycoside hydrolase family 68 protein {ECO:0000313|EMBL:RNB51771.1}
MSTSTRRMRRPLVGGITAAGVLVGTLFTGTAAAVAQPDLQPGAEPTVHTQQAYAPEDDFTAKWTRADAKQLQRLSDPNAASRENSMPASLTMPTVPQDFPDMSNGEVWVWDTWPLTDEHGNQYSVNGQEIIFSLVADRSLGFDDRHVFAKIGYFFRPAGIPADERPENGGWTYGGLVFKEGVTGTIFEDQSYSHQTQWSGSARIMKGGEIKLFFTDVAFYRDAAGNNIKPYDPRIALSVGKVHANKNGVKLTGFDKVTDLLQADGTYYQTGEQNPYFNFRDPFTFEDPAHPGETYMVFEGNSAMPRESAQCTEEDLGYREGDPYAESLDEVNASGATYQIGNIGLAKAKNDELTEWEFLPPILSANCVTDQTERPQFVFKDGKSYLFTISHRGTFAAGLDGPEGVYGFVGDGIRSDFQPLNSGSGLALGNPTNLNFAGGQPFAPDYNQHPGHFQAYSHYVMPGGLVQSFIDTIGTHDDFVRGGTLAPTVKMDIEGASTSVDYSYGTGGLGGWADIPANLDVNSGGKLQ
>A0A3N4A6A9.1/58-241 Glycoside hydrolase family 68 protein {ECO:0000313|EMBL:ROZ64321.1}
MPEDSNLADRQISRRSFGRLAAAVTAAATLTSVSYRTDKAWAAEGPQPTPHTQQAYDPISSDFAAKWTRADARQIMTQQNDESVPRGENSMPDSVTMPEIPKDFPVMNDEVWVWDTWSLTDEHSNQMSYKGWDVIFSLVAPRSIPFDSRHWSAKIGYFYRKSTAELNSSWTYGGLLFPEDSSLSNTEWSGSTRLMQGNKVRVFYTATTFGTPRPGTDGKPAVLATVEGRISASDEGVGGGVLGSEETHVTEEDLGYVDGDPFAETVEEVNGMGAHYQRGNIVLAWADSKDLTSWTMLPPILTAYGVNDQTERPQMILQDGKYYLFTISHQFTYANGLRGPDGVYGFVGDGIRSDFQPLNDSGLVLGSPTDLNLPADAPESPTVGQNPRQFQAYSHYIQPGGLIQSFIDNVDGRRGGTLSPTVKVNISGETTAVDYSFGDNGLGPYAYIPTNKRKGGVGHYK
>A0A3S1DAP8.1/37-483 Glycoside hydrolase family 68 protein {ECO:0000313|EMBL:RUT33589.1}
MNIRKFVRRAAAVTFTTALIAGGGTSAFAKEKDSQDYKESYGFSHITRSDMLKIPGQQLSQQFQAPSFDASTIKNIPSAKGVDQWGNPIDIDVWDSWPLQNADGTVANYHGYHIVFALAGDPKRGWDTFIYMFYQKVGDTSIDSWKNAGRVFKDNDKNVPNDPILNKQAEEWSGSATLTKSGDVRLFYTNRQGWDPEHGFFGKQTLTTAQVNLSKPDADTLKVDGVEDFKSIFDGDGKTYQNVDQAFGGGDYADNHTLRDPHYVEDNGHKYLVFESNTGTETGYQGEASLNNRSYYGGSNVFFQSERNKLLSSPKKSLATLANGSLGIIELNDDYSLKKVMKPLIASNTVTDEIERANIFKMDGKWYLFTDSRGSKMTIDGIGDNDVYMLGYVADSLTGPYKPLNKSGLVLHMGLDPKDLTWTYSHFAVPQAEGNNVVITSYMTNRGLFADHHSTFAPSFLVNIKGSKTSVVKDSILEQGQLTVK
>A0A437SU42.1/117-568 Levansucrase {ECO:0000313|EMBL:RVU70410.1}
MHNSKKHSMLTLVSAGVLLGILNTANCKPVHADTLNSPTTNTNNVQNQNTTDQTQNFSSNDSASVTNADQTSDSTQTSTQASDLQLTDEAKKALADVDIDEYSLNQDQIEKINQINFNDINIDTGTQWTYDQFRGTANKMIDQDKQYRVPYFNAKKIKNMPATYTKDAQTGKKTKLDVWDSWPIQDEKTGHVVNYKGYQLAIAMMGQPKHNDSHIYLLYNKYGDNKLSHWKTAGPIFGFKGTPLDQEWSGSATVNSDGTIQLFYTDVDTRKGDNHQKISTANLVLKIDKKGRLKIAKVRNRHELFEGDGYHYQTYKQWKSTNKGADNIAMRDAHVINVDGDRYLVFEASTGTENYQGEDQIYNWQNYGGSDQENLDDFLKINSDDDMKSRASWANAAIGILRLDDNENNPSVEEVMIPLITSPMVSDEIERPNIVPMNGRYYLFATTRLNHGTNDALWRKANRVVGDNVTMLGWVSDHLTYGYEPLNGDATVLTASVPFNWRTGTYSYYAVPFAGSDDQVLITSYMTNRGYAAGKGKRATWAPSFLLRINPDDTTEVEDVVTNQGDWIYDESSRNDKMLATSLAGARLKGEPDPDVTGELHKIPLKKKHHKKTSKKKAHHSKKHVAKKQHRR
>B0RBJ8.1/47-509 Sugar hydrolase {ECO:0000313|EMBL:CAQ02894.1}
MTKRIRRGLSASAAATLVVASALLAGGSAQAAGTTPPRPTVHTQEAYAPEDDFTAHWTRADAKQIAKLSDPTAAPRQNSMPEALTMPQVPQDFPSMTDQAYVWDTWPLTDSSGQTYSVDGYDVIFALTAPRTLSFDDRHTYAKIGYFTRPTGIPAEQRPENGGWTYQGNVFEDGVTDGIFPDQSFSQQAEWSGSARIMADGTVKLFFTDVAFYRDAKGQDVKPADPVISLSQGRVEKVDGAVALKGFETVTPLLRPDGQKYQTNEQNWSTNFRDPFTFTDPDHPGKTYMVFEANVAGKRGEQRCDAADLGYRKGDPAAEDPKEVTASGANYQMASIGLAVADDADLTKWHYLDPLLESACVTDQTERPEVMIENGKHYLFTISHRSTFANGIDGPEGVYGFVGNGLRSDYKPMNGGSGLVLGNPTNLNYAGGTAYAPDYNQTPGAFQAYSSYILPGGLVESFIDAVGSKESFRRGGTLGPTVRLEFDGDTSELDRGYGKGGLGGYADIPTTRAFDPAHPPQ
>A0A024PB57.1/37-486 Levansucrase {ECO:0000313|EMBL:CDQ25652.1}
MNLKLLAKKATILTLSTAILAGGSGLVHAEQKAHEDTKEDYGISHITRADMEAMAKQHGNDNFEVPKFDASTIQNIPSATKVTENGEEIKLDVWDTWPLQNADGTVAEYQGYHLLFGLAGKPGHPEDTHIYLFYKKADETSIDAWKNAGRVFDENDKFEANDQFLKDQSEEWSGSATFTSDGEIRLFYTNRTEFNVDKEQYGKQTLTTAQVNVSEANNGTIRVDGVEDHKSIFEGQDGSIYQNVMQAFGGETINWKENHTLRDPHYIEEDGHKYLVFEANTGTNYGYSGEESLYNKAFYGNSEKFFRTEKEQLLNSSKKEYAELANGAIGIIEITDDYELKKVMEPLVASNTVTDEIERPNIFKKDGKWYLFTSARGSKMTIDGIDEEDIYLLGYVSNSLTGEFKPMNKTGIVLHHDLDPYDITWNYAHYVIPQEGTDETVVTSYMTNRGYFKDHKSTFAPSFKLELNGKKSSVVENGILDQGQITLHGEE
>A0A031JWY6.1/225-382 Levansucrase {ECO:0000313|EMBL:EZP81424.1}
MNPHPATLESYPTLQWTARDLARLRDISIPRAPVFTPSDVRRIAGNLDVWDAWPLASMHGIPVRWRGGELWFALAAPAFDDPEERHGTARIHHFHRIDGQFRHIGATFEEGFTPGSREWSGSALFEDGVVTLYFTAAGERGETEPTFRQRIFACTTGLAETGDAVFTGWSDPVEQVIPANIVYMNRHDGTGELGEIKAFRDPAPWRAEDGVDHLLFTASSARHLGLHNGLIGCARRGDADPEGGVFEKRPPLVDASGVNNELERPHIVAYDGRLYLFWSTQAKVFAPGIDAPTGLYGAVADGFGGPWRLLNGHGLVFANPQGEPFQAYSWWVLPDLSVTSFVDYWGIDDAAAKRKPEGRAHFGGTFAPFLHLRLEGDVATLR

``

4. The input file is named `fasta_input.txt`. 
5. The program will read it line by line and it will add in a new output file named `output_file_no_rep.txt` all the sequences with no repetitions.

```
input("Press Enter to continue or Ctrl-z to stop")

input_file = open("fasta_input.txt","r")                              # Open the file and read it
line_in_file = input_file.readlines()                                 # To read each the whole file line by line
var_flag=0                                                            # Set the var_flag to 0
                                            
output_file_no_rep = open("output_file_no_rep.txt", "w")              # Creating an empty file
list_a_num = []                                                       # Creating an empty dictionary

for line in line_in_file: 
    
    if var_flag==1:                                                  # This section prints the protein sequences that are below the line that starts with <. How does it do it? In the first loop, since we have set var_flag to 0, the "if var_flag==1" will not print or write anything. Then, the next "if line startswith ">" and it is not duplicated, will print the line and it will set the var_flag to 1, which indicates that the next line should be printed in the next cycle of the for loop. After this second line is printed, we set the var_flag to 0 again, restarting the whole process. 
        print(line)
        output_file_no_rep.write(line) 
        var_flag=0
    
    if line.startswith('>'):                                          # Reading only lines that start with ">""
        var_accession_number=line[line.find(">")+1:line.find(".")+2]  # Creating a variable that contains only the Accession Number
        if var_accession_number not in list_a_num:                    # If the Accession Number is not in the list, put it  
            list_a_num.append(var_accession_number)
            output_file_no_rep.write(line) 
            print (line)
            var_flag=1    

input_file.close()
output_file_no_rep.close()
input("Press Enter to exit")  

```
